src/test/java/net/minecraftforge/debug/EventPistonDebug.java

```diff
@@ -0,0 +1,115 @@
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.text.TextComponentString;
+import net.minecraft.world.World;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.world.PistonEvent;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.fml.common.event.FMLInitializationEvent;
+import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+import net.minecraftforge.fml.common.registry.GameRegistry;
+
+@Mod(modid = EventPistonDebug.MODID)
+public class EventPistonDebug
+{
+    public static final String MODID = "EventPistonDebug";
+    
+    private static Block shiftOnMove = new BlockShiftOnMove();
+    @Mod.EventHandler
+    public void preInit(FMLPreInitializationEvent event)
+    {
+        GameRegistry.register(shiftOnMove);
+        GameRegistry.register(new ItemBlock(shiftOnMove).setRegistryName(shiftOnMove.getRegistryName()));
+    }
+
+    @Mod.EventHandler
+    public void init(FMLInitializationEvent event)
+    {
+        MinecraftForge.EVENT_BUS.register(this);
+    }
+
+    @SubscribeEvent
+    public void onPistonExtend(PistonEvent.PistonExtendEvent event)
+    {
+        if (event.getWorld().isRemote)
+        {
+            BlockPistonStructureHelper pistonHelper =
+                    new BlockPistonStructureHelper(event.getWorld(), event.getPos(), event.getFacing(), true);
+            for (EntityPlayer player : event.getWorld().playerEntities)
+            {
+                if (pistonHelper.canMove())
+                    player.addChatMessage(new TextComponentString(String.format("Piston will extend moving %d blocks and destroy %d blocks",
+                            pistonHelper.getBlocksToMove().size(), pistonHelper.getBlocksToDestroy().size())));
+                else
+                    player.addChatMessage(new TextComponentString("Piston won't extend"));
+            }
+        }
+        else
+        {
+            World world = event.getWorld();
+            BlockPos pushedBlockPos = event.getPos().offset(event.getFacing());
+            if (world.getBlockState(pushedBlockPos).getBlock().equals(shiftOnMove) && event.getFacing() != EnumFacing.DOWN)
+            {
+                world.setBlockToAir(pushedBlockPos);
+                world.setBlockState(pushedBlockPos.up(), shiftOnMove.getDefaultState());
+            }
+        }
+        event.setCanceled(event.getWorld().getBlockState(event.getPos().offset(event.getFacing())).getBlock() == Blocks.COBBLESTONE);
+    }
+
+    @SubscribeEvent
+    public void onPistonRetract(PistonEvent.PistonRetractEvent event)
+    {
+        if (!event.getWorld().isRemote)
+        {
+            for (EntityPlayer player : event.getWorld().playerEntities)
+            {
+                boolean isSticky = event.getWorld().getBlockState(event.getPos()).getValue(BlockPistonExtension.TYPE) == EnumPistonType.STICKY;
+                if (isSticky)
```

Comment:
Calculate `isSticky` outside of the loop can simplify the code and improve performance.